File: 47-permutations-ii/permutations-ii.cpp
class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> temp;
        sort(nums.begin(), nums.end());
        vector<bool> used(nums.size(), false);

        solve(nums, ans, temp, used, 0);

        return ans;
    }

    void solve(vector<int> &nums, vector<vector<int>> &ans, vector<int> &temp, vector<bool> &used, int index){
        if (index == nums.size()){
            ans.push_back(temp);
            return;
        }

        for(int i=0; i<nums.size(); i++){
            if (used[i] || (i > 0 && nums[i] == nums[i-1] && !used[i-1])) continue;

            temp.push_back(nums[i]);
            used[i] = true;
            solve(nums, ans, temp, used, index+1);
            used[i] = false;
            temp.pop_back();
        }
    }
};

File: 47-permutations-ii/permutations-ii.go
func solve(nums []int, idx int, temp []int, ans *[][]int, visited []bool) {
    if idx == len(nums) {
        tmp := make([]int, len(temp))
        copy(tmp, temp)
        *ans = append(*ans, tmp)

        return
    }

    for i := 0; i < len(nums); i++ {
        if visited[i] || (i > 0 && nums[i] == nums[i-1] && !visited[i-1]) {
            continue
        }

        temp = append(temp, nums[i])
        visited[i] = true
        solve(nums, idx+1, temp, ans, visited)
        visited[i] = false
        temp = temp[:len(temp)-1]
    }
}

func permuteUnique(nums []int) [][]int {
    sort.Slice(nums, func(i, j int) bool {
        return nums[i] < nums[j]
    })
    var ans [][]int
    var temp []int
    visited := make([]bool, len(nums))

    solve(nums, 0, temp, &ans, visited)

    return ans
}